### ISSUE TITLE
Support for LoctekMotion/FlexiSpot desks

### ISSUE THREAD

**Issue by Jolanrensen** [tjhorner/upsy-desky#13]
Looks like most of the work is already done:
https://github.com/iMicknl/LoctekMotion_IoT
I think it would be awesome to have some sort of merge between these projects so that a clean all-in-one (no soldering) solution might also exist for Loctek desks.

**Comment by tjhorner** [comment]
Definitely on the roadmap! Will need an adapter cable as noted in #5.

**Comment by Jolanrensen** [comment]
Awesome! So same connector, different pinout. That's a bit more tricky than expected, but at least it's the same connector :)

**Comment by Jolanrensen** [comment]
I ordered one! Order number #351096
If you already have access to an adapter for that desk for me to try, please let me know! Otherwise I'm also willing to wait for the upsy desky until you can ship it together with an adapter. That's fine by me :) let me know how I can be of help

**Comment by Jolanrensen** [comment]
Upsy desky in the house! Now I just need an adapter :) (or a small guide on how to rewire the cables and make an adapter cable)

**Comment by Jolanrensen** [comment]
I'll move my software findings here, so the adapter research (https://github.com/tjhorner/upsy-desky/issues/5) can happen in the right issue. Using the adapter described in that issue and my adapted version from LoctekMotion_Iot: https://github.com/Jolanrensen/LoctekMotion_IoT/tree/main 

I now have the following working: 
- up/down switch (requires virtual screen to be turned on)
- min/max height
- cover up/down/stop
- target height in cm
- 90% of the time smooth motion and no crashes anymore on the upsy desky

I did find that we probably cannot use the upsy desky templates atm, since those assume setting a pin on/off will control the desk, while LoctekMotion (and more) desks require uart streams instead. We could however maybe generalize that to a switch that should be turned on instead since that's how I got it to work. Also, the target height seems to work very accurately :), no +0.4 needed.

Now, getting the screen and buttons to work is next. I'm curious to see what I can do with those. Since the target height works so well from home assistant, I might connect the buttons directly to home assistant as well, allowing them to map to anything you'd like to happen... Don't know yet. The LocktekMotion_Iot project simply captures the signal from the buttons and activates a switch which then sends the signal to the desk again, but the preset buttons don't seem to do anything for me atm, so I'm not sure where those presets are physically saved.

**Comment by Jolanrensen** [comment]
@tjhorner btw, is there a reason you use "return id(desk_height)->get_last_read() > x;" instead of "return id(desk_height).state > x;"? Is that for performance or for something else?

**Comment by tjhorner** [comment]
@Jolanrensen It's so the automation is more responsive. The state is published less often than the height is read.

**Comment by Jolanrensen** [comment]
Additionally fixed in my branch (https://github.com/Jolanrensen/LoctekMotion_IoT/tree/main)
  - Target height in cm now always lands on the right number. Even if it overshoots at first, it will go up/down till it hits the correct height exactly. (Can be quite comical sometimes)
  - Buttons 1, 2, 3 don't work passed through to my desk for some reason. However, you can now set 3 presets cm in Home Assistant (saved in EspHome for when Hass is offline) where the desk will move to
  
Need to decide what to do with M, A buttons. Maybe I'll recreate some sort of save-function.

**Comment by Jolanrensen** [comment]
Pressing M and then 1, 2, or 3 saves the current desk height to the respective preset. Press M again (or any of the other buttons) to cancel.
Unfortunately, the screen doesn't display that you're saving something since the saving functionality of the desk itself doesn't seem to work for me in this setup

**Comment by r14n** [comment]
What's the status of Flexispot support?  There's some decent looking ones on Amazon

**Comment by Jolanrensen** [comment]
@r14n I've been using my [adapter](https://github.com/tjhorner/upsy-desky/issues/5) and [esphome file](https://github.com/Jolanrensen/LoctekMotion_IoT/blob/main/packages/esphome/hs01b-1_upsy_desky.yaml) successfully for a couple of months now. It's not perfect but it works.

**Comment by r14n** [comment]
> @r14n I've been using my [adapter](https://github.com/tjhorner/upsy-desky/issues/5) and [esphome file](https://github.com/Jolanrensen/LoctekMotion_IoT/blob/main/packages/esphome/hs01b-1_upsy_desky.yaml) successfully for a couple of months now. It's not perfect but it works.

Do you think it would work with this one?  Bit hard to tell if the controller is RJ-45.  

https://www.amazon.com/gp/product/B0CHVQ5LCL/ref=ox_sc_act_title_1?smid=A3686AOLDPXXOD&th=1

![image](https://github.com/tjhorner/upsy-desky/assets/4690326/69e7a1af-db90-40f1-8677-7abdb8daa801)

Which desk model do you have btw?

**Comment by Jolanrensen** [comment]
> > @r14n I've been using my [adapter](https://github.com/tjhorner/upsy-desky/issues/5) and [esphome file](https://github.com/Jolanrensen/LoctekMotion_IoT/blob/main/packages/esphome/hs01b-1_upsy_desky.yaml) successfully for a couple of months now. It's not perfect but it works.
> 
> Do you think it would work with this one? Bit hard to tell if the controller is RJ-45.
> 
> https://www.amazon.com/gp/product/B0CHVQ5LCL/ref=ox_sc_act_title_1?smid=A3686AOLDPXXOD&th=1
> 
> ![image](https://user-images.githubusercontent.com/4690326/278362899-69e7a1af-db90-40f1-8677-7abdb8daa801.png)
> 
> Which desk model do you have btw?

That looks like a completely different controller. Probably RJ-45 since most desks are controlled using that, but we can't be 100% certain. Might also require a different adapter compared to mine. I've got a HS01B-1 controller.

**Comment by r14n** [comment]
Are we able to identify this controller?  Searching for your HS01B-1 is not really turning up useful results (as far as desks I can purchase right now)

Keypad：
E2 Pro: 4 Presets Buttons with USB Ports
![image](https://github.com/tjhorner/upsy-desky/assets/4690326/4adf43a1-abf1-4a9e-a20b-6fc13093e48c)

**Comment by Jolanrensen** [comment]
> @Jolanrensen also, are you selling your adapter or willing to? Wishing for an "off the shelf" solution, as I'm familiar with homeassistant, but my electrical/wiring skills are very minimal. Seems like the upsy-desky isn't currently totally plug n play with flexispot. I'm also trying to keep my desk budget < $300 including additional "smart" hardware.

Haha I'm flattered, but my electrical skills aren't sell-worthy, plus I can't test it with your desk to see if it would work. Besides, my skills were minimal too before starting. I opted for the upsy-desky over a general ESP board since it already had a case and RJ45 ports. There are many toolless connectors that allow you to more easily make an adapter like mine, but nonetheless it will require some effort on your side. AFAIK there's no off-the-shelf solution for Flexispot desks.

**Comment by Loic691** [comment]
Hi To follow this thread.
I just bought an E7 flexispot with this screen
A chance to get it work with Upsy ?
What kind of job we have to produce for flexipot compatibility ?
![20240611_231404](https://github.com/tjhorner/upsy-desky/assets/60837526/5db89504-9d27-43ad-87db-df35eb0c5901)
![20240611_231045](https://github.com/tjhorner/upsy-desky/assets/60837526/73ab74ba-3c4c-4d6b-8a75-e8a0c45f298a)
![20240611_231324](https://github.com/tjhorner/upsy-desky/assets/60837526/e17fc571-dcf9-4b4e-a5e8-f627f9c86f17)

**Comment by garymc** [comment]
Would be interested to know if anyone has made this work with Flexispot desks.

**Comment by Ramiel99** [comment]
One more vote for Flexispot. I have E7Q (https://flexispot.co.uk/4-leg-standing-desk-e7q) 

![Image](https://github.com/user-attachments/assets/85b635fb-593c-48a3-82cd-80226f60798f)

![Image](https://github.com/user-attachments/assets/bc6f8041-862c-4302-beee-e1242352e9d2)

![Image](https://github.com/user-attachments/assets/1065f418-0203-4799-8d98-fe774181a556)

**Comment by L-e-c-o** [comment]
Great project, I was wondering if there were any updates on compatibility with Flexispot ?

**Comment by KingOfSpades** [comment]
> Great project, I was wondering if there were any updates on compatibility with Flexispot ?

Wondering the same!
